src/Microsoft.OData.Core/Json/TranscodingWriteStream.cs

```diff
@@ -0,0 +1,438 @@
+﻿//---------------------------------------------------------------------
+// <copyright file="TranscodingWriteStream.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+#if NETCOREAPP3_1_OR_GREATER
+using System;
+using System.Text;
+using System.Threading.Tasks;
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+
+namespace Microsoft.OData.Json
+{
+    /// <summary>
+    /// A stream that allows converting from one encoding to another when writing to a stream.
+    /// </summary>
+    /// <remarks>
+    /// This is adapated from the .NET runtime's TranscodingStream available in .NET 5+
+    /// see: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Private.CoreLib/src/System/Text/TranscodingStream.cs
+    /// </remarks>
+    internal sealed class TranscodingWriteStream : Stream
+    {
+        // We optimistically assume 1 byte ~ 1 char during transcoding. This is a good rule of thumb
+        // but isn't always appropriate: transcoding between single-byte and multi-byte encodings
+        // will violate this, as will any invalid data fixups performed by the transcoder itself.
+        // To account for these unknowns we have a minimum scratch buffer size we use during the
+        // transcoding process. This should be generous enough to account for even the largest
+        // fallback mechanism we're likely to see in the real world.
+
+        private const int MinWriteRentedArraySize = 4 * 1024;
+        private const int MaxWriteRentedArraySize = 1024 * 1024;
+
+        private static ValueTask CompletedValueTask => default;
+
+        private readonly Encoding _innerEncoding;
+        private readonly Encoding _thisEncoding;
+        private Stream _innerStream; // null if the wrapper has been disposed
+        private readonly bool _leaveOpen;
+
+        /*
+            * Fields used for writing bytes [this] -> chars -> bytes [inner]
+            * Lazily initialized the first time we need to write
+            */
+
+        private Encoder _innerEncoder;
+        private Decoder _thisDecoder;
+
+        internal TranscodingWriteStream(Stream targetStream, Encoding targetEncoding, Encoding inputEncoding, bool leaveOpen = false)
+        {
+            Debug.Assert(targetStream != null);

```

Comment:
Should we use `Debug` assertions here or throw exceptions?

Reply:
Since this is internal and used in controlled scenarios, I didn't think it was necessary to throw the exception.